### PR TITLE
[add] 16 directional augmentation for foa format

### PR DESCRIPTION
**Pull request by lsg1213**

### CODE REVIEW

File: transforms.py [hunk 1]
```diff
@@ -47,9 +47,22 @@ def foa_intensity_vec_aug(x, y):
     flip = tf.cast(flip, 'float32')
 
     intensity_vectors = (1 - 2*tf.reshape(flip, (-1, 1, 1, 3))) * intensity_vectors 
-    cartesian = (1 - 2*tf.reshape(flip, (-1, 1, 3, 1))) * cartesian 
-
-    x = tf.concat([x[..., :-3], intensity_vectors], axis=-1)
+    cartesian = (1 - 2*tf.reshape(flip, (-1, 1, 3, 1))) * cartesian
+
+    # x,y축 회전
+    correct_shape = tf.reshape(tf.tile([0,1,2], [tf.shape(x)[0]]), (tf.shape(x)[0], 3))
```
Comment (daniel03c1) [inline transforms.py]: tf.shape(x)[0]가 너무 많이 나오는데, batch_size 등과 같이 이름을 붙여서 통일시키는건 어떤가요?

그리고 tf.reshape(...)이거를 tf.tile([[0, 1, 2]], [tf.shape(x)[0], 3]) 이렇게 바꾸면 더 간단하지 않을까 생각이 듭니다.

File: transforms.py [hunk 2]
```diff
@@ -47,9 +47,22 @@ def foa_intensity_vec_aug(x, y):
...
+    current_shape = tf.reshape(tf.tile([0,2], [tf.shape(x)[0]]), (tf.shape(x)[0], 2))
+    perm = tf.map_fn(tf.random.shuffle, current_shape)
+    perm = tf.gather(tf.concat([perm, tf.ones((tf.shape(x)[0],1), dtype=perm.dtype)],-1), [0,2,1], axis=-1)
+    check = tf.reduce_sum(tf.cast(perm != correct_shape, tf.int32), -1)[..., tf.newaxis]
```
Comment (daniel03c1) [inline transforms.py]: [..., tf.newaxis]도 좋은데 tf.reduce_sum(..., -1)[..., tf.newaxis]에서 tf.reduce_sum(..., -1, keepdims=True)로 바꾸면 
어떨까요?

File: transforms.py [hunk 3]
```diff
@@ -47,9 +47,22 @@ def foa_intensity_vec_aug(x, y):
...
-    cartesian = (1 - 2*tf.reshape(flip, (-1, 1, 3, 1))) * cartesian 
-
-    x = tf.concat([x[..., :-3], intensity_vectors], axis=-1)
+    cartesian = (1 - 2*tf.reshape(flip, (-1, 1, 3, 1))) * cartesian
+
+    # x,y축 회전
+    correct_shape = tf.reshape(tf.tile([0,1,2], [tf.shape(x)[0]]), (tf.shape(x)[0], 3))
+    current_shape = tf.reshape(tf.tile([0,2], [tf.shape(x)[0]]), (tf.shape(x)[0], 2))
+    perm = tf.map_fn(tf.random.shuffle, current_shape)
+    perm = tf.gather(tf.concat([perm, tf.ones((tf.shape(x)[0],1), dtype=perm.dtype)],-1), [0,2,1], axis=-1)
```
Comment (daniel03c1) [inline transforms.py]: 지금 이 방법도 좋은데요 각 샘플별로 첫 숫자만 뽑으면 되니까 그런 방법은 어떤가요?
perm = 2 * tf.random.uniform([batch_size, 1], maxval=2, dtype=tf.int32) 
perm = tf.concat([perm, tf.ones_like(perm), 2-perm], axis=-1)
로 하면 조금 더 간단해지지 않을까 생각해봤습니다.

File: transforms.py [hunk 4]
```diff
@@ -47,9 +47,22 @@ def foa_intensity_vec_aug(x, y):
...
+    check = tf.reduce_sum(tf.cast(perm != correct_shape, tf.int32), -1)[..., tf.newaxis]
+    feat_perm = (perm + check) % 3
+
+    perm = tf.gather(tf.concat([tf.random.shuffle([0,2]), [1]],-1), [0,2,1])
```
Comment (daniel03c1) [inline transforms.py]: 위에서 sample별로 [0, 1, 2], [2, 1, 0]을 뽑아서 feat를 뽑고, 거기를 기반으로 feat_perm을 했는데
또 한번 feat_perm과 무관하게 배치별로 뽑는 것 같습니다.
실제 H1, H2, H3는 배치별 perm을 적용하면서, intensity vector, y는 sample별 perm을 적용하는 것 같습니다.
확인 부탁드립니다!

File: transforms.py [hunk 5]
```diff
@@ -60,22 +73,3 @@ def split_total_labels_to_sed_doa(x, y):
     n_classes = tf.shape(y)[-1] // 4
     return x, (y[..., :n_classes], y[..., n_classes:])
 
-
-''' For FOA Channel Swapping '''
-T_XYZ_TO_YZX = tf.convert_to_tensor(
```
Comment (daniel03c1) [inline transforms.py:65]: 이 함수 없애신 만큼 test파일에서도 같이 삭제 진행하면 좋을 것 같습니다.

Comment (lsg1213) [reply transforms.py:65]: 위에꺼 테스트 하는 와중에 test도 같이 지웠습니다

Comment (daniel03c1) [reply transforms.py:65]: 저한테는 config랑 transforms.py 두가지만 수정된 것으로 나오거든요
혹시 지워진 것은 어디서 확인할 수 있는지 알 수 있을까요?

File: transforms.py (hunk 4 above)
Comment (lsg1213) [reply transforms.py]: 아 xyz축 코드랑 섞여있다가 지우는 과정에서 손을 잘못봤네요 수정하겠습니다

Comment (daniel03c1) [reply transforms.py]: 알겠습니다 혹시 해당 코드가 이전에 xy, xyz, xz? 실험할 때는 없었기 때문에
별도로 다시 돌려서 테스트해볼 필요는 없는거죠?

File: transforms.py (hunk 5 above)
Comment (lsg1213) [reply transforms.py:65]: 아직 push를 안 했어요 해당하는 test 코드까지 다 지워서 같이 push하겠습니다

File: transforms.py (hunk 4 above)
Comment (lsg1213) [reply transforms.py:60]: 오늘 오전에 진행했던 테스트는 약간 께름칙해서 다시 돌려보려고요

File: transforms.py (hunk 1 above)
Comment (lsg1213) [reply transforms.py]: batch_size로 받아서 사용하는 것으로 바꿨습니다. 

tf.tile로 다 해결보는 건 좋은 아이디어 같습니다 다만 아래와 같이 [~, 1]로 사용해주어야 되는 것 같아요
tf.tile([[0, 1, 2]], [tf.shape(x)[0], 1]) 

수정했습니다.

File: transforms.py (hunk 2 above)
Comment (lsg1213) [reply transforms.py]: 감사합니다.

File: transforms.py (hunk 3 above)
Comment (lsg1213) [reply transforms.py]: O(n)으로 같긴한데 확실히 연산량은 더 적긴 한 것 같긴합니다. 제안해주신 코드 감사합니다.

**Review by daniel03c1 — APPROVED** [review]
좋은 것 같습니다

File: transforms.py [hunk 6]
```diff
@@ -37,19 +37,36 @@ def foa_intensity_vec_aug(x, y):
...
+    # perm = tf.map_fn(tf.random.shuffle, correct_shape)
+    
+    check = tf.reduce_sum(tf.cast(perm != correct_shape, tf.int32), -1, keepdims=True)
+    feat_perm = (perm + check) % 3
+
+    intensity_vectors = tf.gather(intensity_vectors, feat_perm, axis=-1, batch_dims=1)
+    cartesian = tf.gather(cartesian, feat_perm, axis=-2, batch_dims=1)
+    
+    x = tf.concat([x[..., :1], tf.gather(x[..., 1:4], perm, axis=-1, batch_dims=1), intensity_vectors], axis=-1)
+
```
Comment (ParkJinHyeock) [inline transforms.py:69]: 이 부분은 잘모르는데 하나하나 코드 처가면서 보니 
intensity의 경우는 x, y, z 
그 나머지는 y,  z, x인 경우라면 잘 작동하는 것 같습니다.

그런데 궁금한게 

    check = tf.reduce_sum(tf.cast(perm != correct_shape, tf.int32), -1, keepdims=True)
    feat_perm = (perm + check) % 3

이 부분의 경우 축이 두개만 바뀌면 성립하는데 축이 세 개바뀌면 성립 안하지 않나요? 계속 두개만 바꿀거라면 괜찮아 보입니다. 유도를 되게 잘하신것 같습니다!

**Review by ParkJinHyeock — APPROVED** [review]

File: transforms.py (hunk 6 above)
Comment (lsg1213) [reply transforms.py:69]: 축이 3개 바뀌는 경우는 h1,2,3과 x,y,z의 transform 숫자가 같은 경우인데 그런 케이스는 perm != correct_shape을 하고 sum을 하면 3이 나옵니다. 그래서 3 % 3은 0이므로 변화가 h1,2,3에 적용되는 transform 숫자와 x,y,z의 숫자가 동일하기 때문에 성립이 된다고 생각했습니다. 혹시 아니라면 안 되는 케이스를 하나 알려주시면 감사드리겠습니다.

Comment (ParkJinHyeock) [reply transforms.py:69]: 저 혹시 h1 h2 h3가 어떤 의미인지 알 수 있을까요? 저는 feat_perm은 xyz의 변환 perm은 yzx의 변환이라고 생각했는데
만약 xyz -> yzx 로의 축변환이 feat_perm이라면 yzx는 zxy가 돼야하지않나요?

Comment (lsg1213) [reply transforms.py:69]: 그 부분은 저도 헷갈린거긴 한데 제가 따로 그림을 드리겠습니다.
